fixes #3763 - extend rabl templates

timogoebel opened the pull request:
This should be considered a proposal on how it would be possible for plugins to extend a rabl template to add more attributes to APIv2 responses.

This is still missing tests and inline documentation. I'd love to hear some feedback on this first.

It can be used like this:

lib/foreman_expire_hosts/engine.rb
```ruby
Foreman::Plugin.register :foreman_expire_hosts do
  [...]
  extend_rabl_template 'api/v2/hosts/main', 'api/v2/hosts/expiration'
end
```

app/views/api/v2/hosts/expiration.json.rabl:
```ruby
attribute :expired_on
```

timogoebel committed:
fixes #3763 - extend rabl templates

ohadlevy commented:
:+1:  overall, lets just be careful not to crate N+1 issues, AFAIR @ShimShtein already solved a similar problem before with the ability to enhanced controller eager loading?

iNecas commented on config/initializers/rabl_init.rb:
```diff
@@ -35,9 +35,37 @@ def data_name(data_token)
+    def call(template)
+      source = template.source
+
+      %{ ::Rabl::Engine.new(#{source.inspect}, :template => '#{template.virtual_path}').
```
Why do we need this?

timogoebel replied:
... to pass the template name (`template.virtual_path`) to the Engine so we can use it in `template_extensions_from_plugins `. An alternative to this would be `@_options[:template] || context_scope.virtual_path`.

iNecas replied:
Any downside of the non-monkey patching version? That would be the option of my choice

timogoebel replied:
I agree that this monkey patching is not ideal, but I am not sure if `context_scope.virtual_path` is always the same as the template name. I think, it just refers to the controller + controller action. That is equal to the template name in most cases, but not always. Or am I missing something here?

iNecas replied:
I see now why this is needed. Thanks for explanation

ShimShtein commented:
Indeed, N+1 is already [solved for hosts](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L23). BTW, facets are added to RABL automatically.

ShimShtein commented on config/initializers/rabl_init.rb:
```diff
@@ -35,9 +35,37 @@ def data_name(data_token)
   end
 end
 Rabl::Engine.send(:prepend, Foreman::RablEngineExt)
+ActionView::Template::Handlers::Rabl.singleton_class.send(:prepend, Foreman::RablTemplateHandlerExt)
```
Did you consider using RABL's `partial` call as I do [with facets](https://github.com/theforeman/foreman/blob/4d563b215b72081bb7b63f292bd297438192e462/app/views/api/v2/hosts/main.json.rabl#L56)?

timogoebel replied:
I believe we would get similar results. What do you think could be gained from this?

ShimShtein replied:
Easier to read, less monkey patching?

timogoebel replied:
Ok, so you are proposing to explicitly add extension points to every template. I thought you were talking about `extends` vs. `partial`. I prefer this approach as it's way more flexible.

ShimShtein replied:
I see that we already overload the `Engine` class anyway.

ShimShtein commented:
Maybe we should move facets to use this framework too, if it's already there - less explicit code in host's rabl.

I mean changing the [facet definition](https://github.com/theforeman/foreman/blob/a4bf52491fe5eeceb8606a540e25bc9ceea8a7ba/app/services/facets/entry.rb#L51) to:
``` ruby
    def api_view(view_templates)
      extend_api_template 'api/v2/hosts/main', view_templates[:single]
      extend_api_template 'api/v2/hosts/list', view_templates[:list]
    end
```

timogoebel commented:
@ShimShtein: Sure, we can do that. But I'd prefer to do this in another PR. It's just a nice to have imho.

All right, I'll take this as a go. Thanks for the fast replies, appreciate it. I'll continue working on this and will let you know once this is ready for a full review.

timogoebel commented:
@ShimShtein, @iNecas et all:

* Renamed the plugin extension point
* Added some inline code comments
* Added tests

Happy to hear more comments. If we could get this into 1.17 this would be awesome for plugin authors. :-)

ShimShtein commented:
Thanks @timogoebel!

Could you please update the [wiki page](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin) accordingly?
